File: backend/src/main/java/com/bigshen/blog/service/NotificationService.java
package com.bigshen.blog.service;

import com.bigshen.blog.dto.NotificationDTO;
import com.bigshen.blog.model.Notification;
import com.bigshen.blog.repository.NotificationRepository;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.PageRequest;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.Optional;

/**
 * 站内通知服务
 */
@Slf4j
@Service
@RequiredArgsConstructor
public class NotificationService {

    private final NotificationRepository notificationRepository;

    /**
     * 获取用户通知列表（分页）
     */
    public Page<NotificationDTO> getNotifications(Long userId, int page, int size) {
        return notificationRepository.findByUserIdOrderByCreatedAtDesc(
                userId, PageRequest.of(page, size))
                .map(this::toDTO);
    }

    /**
     * 获取未读通知数量
     */
    public long getUnreadCount(Long userId) {
        return notificationRepository.countByUserIdAndIsReadFalse(userId);
    }

    /**
     * 标记单条通知为已读
     */
    @Transactional
    public void markAsRead(Long notificationId) {
        Optional<Notification> optional = notificationRepository.findById(notificationId);
        optional.ifPresent(notification -> {
            notification.setIsRead(true);
            notificationRepository.save(notification);
        });
    }

    /**
     * 标记所有通知为已读
     */
    @Transactional
    public void markAllAsRead(Long userId) {
        notificationRepository.markAllAsRead(userId);
    }

    /**
     * 创建通知（独立事务，失败不中断主流程）
     */
    @Transactional(propagation = org.springframework.transaction.annotation.Propagation.REQUIRES_NEW)
    public void createNotification(Long userId, String type, String message, Long relatedId) {
        try {
            Notification notification = Notification.builder()
                    .userId(userId)
                    .type(type)
                    .message(message)
                    .relatedId(relatedId)
                    .isRead(false)
                    .build();
            notificationRepository.save(notification);
            log.info("通知已创建: userId={}, type={}, message={}", userId, type, message);
        } catch (Exception e) {
            // 通知创建失败不应影响主业务（评论/点赞）
            log.error("通知创建失败（不影响主业务）: userId={}, type={}, error={}",
                    userId, type, e.getMessage());
        }
    }

    /**
     * 转换为 DTO
     */
    private NotificationDTO toDTO(Notification notification) {
        return NotificationDTO.builder()
                .id(notification.getId())
                .type(notification.getType())
                .message(notification.getMessage())
                .relatedId(notification.getRelatedId())
                .isRead(notification.getIsRead())
                .createdAt(notification.getCreatedAt())
                .build();
    }
}

File: frontend/src/services/notificationService.ts
import api from './api';

export interface NotificationItem {
    id: number;
    type: string;
    message: string;
    relatedId: number;
    isRead: boolean;
    createdAt: string;
}

/**
 * 通知相关 API
 */
export const notificationService = {
    /** 获取通知列表 */
    getNotifications: async (page = 0, size = 20): Promise<{
        content: NotificationItem[];
        totalPages: number;
        totalElements: number;
    }> => {
        const response = await api.get('/notifications', {
            params: { page, size },
        });
        return response.data;
    },

    /** 获取未读通知数 */
    getUnreadCount: async (): Promise<number> => {
        const response = await api.get<{ count: number }>('/notifications/unread-count');
        return response.data.count;
    },

    /** 标记单条已读 */
    markAsRead: async (notificationId: number): Promise<void> => {
        await api.put(`/notifications/${notificationId}/read`);
    },

    /** 标记全部已读 */
    markAllAsRead: async (): Promise<void> => {
        await api.put('/notifications/read-all');
    },
};
